### PR TITLE
docs(blog): network deepdive

### PR DESCRIPTION
**Pull request by RossMurr4y**
## Description

Blog article exploring the network component in depth.

## Target Audience

- [x] hamlet users
- [ ] hamlet framework developers (incl. plugins)

#### Documentation
- [ ] Spelling / Syntax correction only
- [ ] Refactor (Documentation overhaul, or revising after changes to hamlet)
- [ ] New feature (Documenting something new)
#### Site Design
- [ ] Refactor (No new or removed content.)
- [x] New Features & Content
- [ ] Docusaurus / Plugin version update.

## Followup Actions

- [x] None

## Checklist:

- [x] I have tested the changes locally - see [README.md](https://github.com/hamlet-io/docs/blob/master/README.md)
- [x] I have added appropriate labels to this PR.
- [ ] I have added this to the `hamlet roadmap` project.

### CODE REVIEW

**Comment by RossMurr4y** [comment]
Drafted PR to discuss best examples to use.

File: blog/2020-07-07-hamlet-network-component.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+The other subnet is private 
+> Create firewall and security group rules required to prevent public access to this subnets.
+
+A deployment-unit in the private subnet is linked to a deployment-unit in one of the public subnets
+> Evaluate the linked component and determine the configuration required for communication with that linked deployment-unit. Add firewall/security group rules to allow communication across subnets between those two components.
+
+#### A note on Subnets and Availability Zones
+
+`Availability Zones`  (AZ’s) are a common concept across cloud providers. They are defined as unique physical locations within a given region. A cloud provider will typically build multiple AZ’s to enable deployment to a single region whilst still preventing putting customer data all in a single data warehouse. In the event of a natural disaster in a given region, the data for that region can be spread out across multiple data centres so there is far less chance for impacting customers.
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:40]: single data warehouse - single data centre ( data wharehouses are a different conpcet )

File: blog/2020-07-07-hamlet-network-component.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+Though it’s a commonality across cloud providers, the implementation of AZ’s differ. On resources that offer AZ’s for AWS, the consumer is asked to define how many AZ’s they would like to distribute their data across with a slight premium for the privilege of the increased reliability. In Azure, some services are offered in the same way as AWS (which they call “Zonal services”) and others are automatically backed up across multiple zones for you and require no additional configuration (known as Zone-redundant services”). So for the above example, we don’t require different availability zones and subsequently use less subnets. For the same implementation in AWS however, we might opt to define additional AZ’s in the solution which would in turn result in more subnets - thankfully Hamlet will do all that for you.
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:42]: `Though its common across cloud providers` ?

File: blog/2020-07-07-hamlet-network-component.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+Two of our subnets are publicly accessible
+> Create all applicable firewall and security group rules required to ensure this.
+
+The other subnet is private 
+> Create firewall and security group rules required to prevent public access to this subnets.
+
+A deployment-unit in the private subnet is linked to a deployment-unit in one of the public subnets
+> Evaluate the linked component and determine the configuration required for communication with that linked deployment-unit. Add firewall/security group rules to allow communication across subnets between those two components.
+
+#### A note on Subnets and Availability Zones
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:38]: Should this be a admonition as a note? Might keep the flow a bit more around the document

File: blog/2020-07-07-hamlet-network-component.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+For the `network` component, a `networkacl` sub-component represents high-level access control of the network, based on 4-tuple rules. 
+
+For those less familiar with 4 tuples they consist of:
+	* Source Port
+	* Source IP address
+	* Destination Port
+	* Destination IP address
+
+By assigning a Network ACL sub-component to a `network` component we can easily define the security of our entire network.
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:68]: `They act as a last line of defence for network security, we apply fine grained rules in security groups which are managed with links in hamlet. Network ACL's are useful for deny lists if you have known bad IP addresses or want to lock your whole network away from particular sources`

Maybe a little more to add around that

File: blog/2020-07-07-hamlet-network-component.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+### IP Address Groups
+
+Before we take a look at what this may look like in a Hamlet solution, Hamlet adds one more utility to our tool belt - IP Address Groups.
+
+In Hamlet, an IP Address Group is a commonly understood IP address range that can be referenced under a commonly understood name. For instance, “_localnet” can be used in place of the CIDR range of the network of a given component, “_global” really does mean global access, and “_segment” can be used to reference a specific Availability Zone (in relation to the component it is defined on). This is considerably easier to maintain and far easier to understand.
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:74]: The `_` at the start of the group identifies hamlet in build IP address groups. They perform specific lookups on various parts of your deployed hamlet. For Instance .... 

You can also define your own IP address groups for common IP address ranges, say to limit access from your proxy servers.  You can define this gorup across your whole hamlet deployment and reference it in all solutions.

File: blog/2020-07-07-hamlet-network-component.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+## Route Tables
+The `network` component has a second sub-component - this time around its for defining specific routes for your network.
+
+The `routetable`  sub-component provides a list of Gateways and conditions that can be used to forward traffic outside of the virtual network. Multiple `routetable` sub-components can be added to a network to allow subnets to use different routing rules. This grants finely tuned control of the flow of traffic in, out and within a network so that only subnets containing resources that need to communicate have access to do so, whilst others can be tightly controlled.
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:79]: The routetable component itself doesn't actually have any routes in it. The gateways that link to it provide routes to the table. 

This allows the gateway to  really control how routing behaves and makes the gateways easy to plug and play with the network access

File: blog/2020-07-07-hamlet-network-component.md (hunk 1 above)
Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md:40]: Good catch, thanks

File: blog/2020-07-07-hamlet-network-component.md (hunk 3 above)
Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md:38]: Agreed, I shall update this one.

File: blog/2020-07-07-hamlet-network-component.md (hunk 2 above)
Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md]: updated

File: blog/2020-07-07-hamlet-network-component.md (hunk 4 above)
Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md]: Great, thanks!

File: blog/2020-07-07-hamlet-network-component.md (hunk 5 above)
Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md]: Added this in too :)

File: blog/2020-07-07-hamlet-network-component.md (hunk 6 above)
Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md]: Added :)

**Comment by RossMurr4y** [comment]
@roleyfoley I've factored in your comments and changes and added an example for the remainder of the content. Marking as ready for review.

File: blog/2020-07-07-hamlet-network-component.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+A `network` deployment-unit exists
+> 3 tiers in our Solution, each with a single-availability zone (more on this in a moment) - total subnets is therefor 3 x 1 = 3. Divide our network CIDR’s range into 3 equal ranges.
+
+Two of our subnets are publicly accessible
+> Create all applicable firewall and security group rules required to ensure this.
+
+The other subnet is private 
+> Create firewall and security group rules required to prevent public access to this subnets.
+
+A deployment-unit in the private subnet is linked to a deployment-unit in one of the public subnets
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:35]: component not deployment unit

File: blog/2020-07-07-hamlet-network-component.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	* Destination Port
+	* Destination IP address
+
+By assigning a Network ACL sub-component to a `network` component we can easily define the security of our entire network. They act as a last line of defence for network security, we apply fine grained rules in security groups which are managed with links in hamlet. Network ACL's are useful for deny-lists if you have known bad IP addresses or want to lock your whole network away from particular sources.
+
+### IP Address Groups
+
+Before we take a look at what this may look like in a Hamlet solution, Hamlet adds one more utility to our tool belt - IP Address Groups.
+
+The `_` at the start of the group name identifies hamlet in build IP address groups. They peform specific lookups on various parts of your deployed hamlet. For instance “_localnet” can be used in place of the CIDR range of the network of a given component, “_global” really does mean global access, and “_segment” can be used to reference a specific Availability Zone (in relation to the component it is defined on). This is considerably easier to maintain and far easier to understand.
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:71]: `in build` - `inbuilt`

File: blog/2020-07-07-hamlet-network-component.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	* Source Port
+	* Source IP address
+	* Destination Port
+	* Destination IP address
+
+By assigning a Network ACL sub-component to a `network` component we can easily define the security of our entire network. They act as a last line of defence for network security, we apply fine grained rules in security groups which are managed with links in hamlet. Network ACL's are useful for deny-lists if you have known bad IP addresses or want to lock your whole network away from particular sources.
+
+### IP Address Groups
+
+Before we take a look at what this may look like in a Hamlet solution, Hamlet adds one more utility to our tool belt - IP Address Groups.
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:69]: I can't add a utility to a tool belt.... Nit pick but want to get things clear.

File: blog/2020-07-07-hamlet-network-component.md (hunk 8 above)
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:71]: Segment is a bit of a weird one, it actually represents the public IP address assigned to outbound calls from the segment. Its the NAT IP address. 

The new `_tier:` would represent a given subnet across availability zones

File: blog/2020-07-07-hamlet-network-component.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+```
+Defined by the above example:
+* An IPAddressGroup called "_exampleIPAddressGroup"
```
Comment (roleyfoley) [inline blog/2020-07-07-hamlet-network-component.md:156]: This example should be `exampleIPAddressGroup` we say above that `_*` address groups are inbuilt ones. This isn't an inbuilt group

Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md]: Cheers, tidied that up!

File: blog/2020-07-07-hamlet-network-component.md (hunk 8 above)
Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md]: fantastic, thanks!

File: blog/2020-07-07-hamlet-network-component.md (hunk 9 above)
Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md]: updated to "Hamlet adds one more feature to our toolkit - IP Address Groups"

File: blog/2020-07-07-hamlet-network-component.md (hunk 7 above)
Comment (RossMurr4y) [reply blog/2020-07-07-hamlet-network-component.md]: ta

**Pull request merged**